### ISSUE TITLE
CTC loss with dynamic length

### ISSUE THREAD

**Issue by githubharald** [tensorflow/tensorflow#12215]
This is a very specific questions, I'm afraid nobody on stackoverflow will ever answer this. I will copy the text from there, the original question can be found there: 
https://stackoverflow.com/questions/45568266/tensorflow-ctc-loss-ctc-merge-repeated-parameter

And I'm not 100% sure if this behaviour is wanted or a bug (I think its the former one but I'm really not sure).

I'm using Tensorflow 1.0 and its CTC loss [1]. When training, I sometimes get the "No valid path found." warning (which harms learning). It is not due to a high learning rate as sometimes reported by other Tensorflow users.

After analyzing it a bit, I found the pattern that causes this warning:
 - feeding an input sequence into the ctc_loss with length seqLen
 - feeding a label with labelLen characters
 - label has numRepeatedChars repeated chars in it, where I count "ab" as 0, "aa" as 1, "aaa" as 2 and so on 
 - warning occurs, when: seqLen - labelLen < numRepeatedChars

Three examples:

 - Ex.1: label="abb", len(label)=3, len(inputSequence)=3 => (3-3=0)<1 is true --> warning
 - Ex.2: label="abb", len(label)=3, len(inputSequence)=4 => (4-3=1)<1 is false --> no warning
 - Ex.3: label="bbb", len(label)=3, len(inputSequence)=4 => (4-3=1)<2 is true --> warning

When I now set the ctc_loss parameter ctc_merge_repeated=False, then the warning disappears.

Three questions:

 - Q1: why is there a warning when repeated chars occur? I thought, as long as the input sequence is not shorter than the target labelling, there is no problem. And when repeated chars are merged in the label, then it gets even shorter, therefore the condition that the input sequence is not shorter still holds.
 - Q2: why does the ctc_loss in its default settings produce this warning? Repeated chars are common in the domains CTCs are used such as handwritten text recognition (HTR)
 - Q3: what settings should I use when doing HTR? Of course labels can have repeated chars. Therefore ctc_merge_repeated=False would make sense. Any suggestions?

Python program to reproduce warning:

```
import tensorflow as tf
import numpy as np

def createGraph():
    tinputs=tf.placeholder(tf.float32, [100, 1, 65]) # max 100 time steps, 1 batch element, 64+1 classes
    tlabels=tf.SparseTensor(tf.placeholder(tf.int64, shape=[None,2]) , tf.placeholder(tf.int32,[None]), tf.placeholder(tf.int64,[2])) # labels
    tseqLen=tf.placeholder(tf.int32, [None]) # list of sequence length in batch
    tloss=tf.reduce_mean(tf.nn.ctc_loss(labels=tlabels, inputs=tinputs, sequence_length=tseqLen, ctc_merge_repeated=True)) # ctc loss
    return (tinputs, tlabels, tseqLen, tloss)

def getNextBatch(nc): # next batch with given number of chars in label
    indices=[[0,i] for i in range(nc)]
    values=[i%65 for i in range(nc)]
    values[0]=0
    values[1]=0 # TODO: (un)comment this to trigger warning
    shape=[1, nc]
    labels=tf.SparseTensorValue(indices, values, shape)
    seqLen=[nc]
    inputs=np.random.rand(100, 1, 65)
    return (labels, inputs, seqLen) 


(tinputs, tlabels, tseqLen, tloss)=createGraph()

sess=tf.Session()
sess.run(tf.global_variables_initializer())

nc=3 # number of chars in label
print('next batch with 1 element has label len='+str(nc))
(labels, inputs, seqLen)=getNextBatch(nc)
res=sess.run([tloss], { tlabels: labels, tinputs:inputs, tseqLen:seqLen } )
```

This is the C++ Tensorflow code [2] where the warning comes from:
```
// It is possible that no valid path is found if the activations for the
// targets are zero.
if (log_p_z_x == kLogZero) {
    LOG(WARNING) << "No valid path found.";
    dy_b = y;
    return;
}
```

[1] https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/nn/ctc_loss
[2] https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/ctc/ctc_loss_calculator.cc

-----

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: costum
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.0.0-rc2-15-g47bba63-dirty, 1.0.0
- **Python version**: 3.5.2
- **Bazel version (if compiling from source)**: -
- **CUDA/cuDNN version**: [problem also on CPU]
- **GPU model and memory**: [problem also on CPU]
- **Exact command to reproduce**: see code above

**Comment by vcarbune** [comment]
Hi,

The warning message you're seeing, as mentioned in the comment in the code, can appear also in cases when ctc_merge_repeated=False.

It appears whenever it's impossible to determine a valid path for the target labeling based on the network activation through the forward/backward variables (which is where the ctc_merge_repeated is applied - note: the code has some references to the Alex Graves' thesis, which I find useful going through).

Q1: why is there a warning when repeated chars occur? I thought, as long as the input sequence is not shorter than the target labelling, there is no problem. And when repeated chars are merged in the label, then it gets even shorter, therefore the condition that the input sequence is not shorter still holds.

Again, this warning changes as training progress (assuming not all the samples in your training set raise this problem), since the network activations change. The warning states that there couldn't be a valid path found based on the forward/backward computation, but not that that the sequence is invalid.

Note that ctc_merge_repeated is different from collapsing repeated labels, which is done through preprocess_collapse_repeated. The former controls how the CTC loss penalizes internally repeated labels ('a' 'a' 'b' becomes 'blank' 'a' 'blank' 'a' 'blank' 'b' 'blank' for the forward/backward computation), while the latter actually modifies the ground truth. ('a' 'a' 'b' first becomes 'a' 'b' and then 'blank' 'a' 'blank' 'b' 'blank' for the forward/backward computation).

Q2: why does the ctc_loss in its default settings produce this warning? Repeated chars are common in the domains CTCs are used such as handwritten text recognition (HTR)

This should be the default CTC behavior, where collapsing labels for the loss/gradient computation allows focusing on learning the label transitions instead of where the label appears.

Q3: what settings should I use when doing HTR? Of course labels can have repeated chars. Therefore ctc_merge_repeated=False would make sense. Any suggestions?

Having ctc_merge_repeated=True does allow learning repeated characters. Please see the detailed comments in the documentation, specifically about the behavior of the different combination of flags "Regarding the arguments..." 

(https://www.tensorflow.org/versions/r0.12/api_docs/python/nn/connectionist_temporal_classification__ctc_)

**Comment by githubharald** [comment]
Hi Victor, thanks for your answer.

The warning always appears on those samples on which the condition (see first post) holds. Even after training.
In the test program, the input is a random input, but even when repeating the code for multiple random inputs, the warning always occurs.
The only way to toggle the warning is by changing the target labelling ... so it appears to me that it has something to to with this target labelling!?

I further decreased the tensor sizes of the test program to have a closer look at a simple example - code see below.
The input has shape T=5 (timesteps) and C=6 (classes).
The target labelling is [0, 0, 2]. The probability for a path p which yields the labelling [0,0,2] is not zero.
E.g. the path 0->blank->0->2->blank (with blank=C-1=5) yields the target labelling. And so do some more paths. 
Therefore, the probability of seeing the target labelling given the input should not be zero.

**Output of test program:**
input:  
[[[ 0.59413256  0.23575521  0.95215654  0.46271613  0.08859916  0.43706737]]

 [[ 0.11860785  0.11229196  0.01522826  0.96020464  0.34278502  0.27091574]]

 [[ 0.29061551  0.27465308  0.94130146  0.12166673  0.86577444  0.37988687]]

 [[ 0.77218394  0.07898683  0.85200237  0.4295197   0.76858367  0.72602145]]

 [[ 0.09071068  0.43796645  0.6265344   0.48837476  0.24057374  0.03459447]]]
 
label:  SparseTensorValue(indices=[[0, 0], [0, 1], [0, 2]], values=[0, 0, 2], dense_shape=[1, 3])

loss:  [inf]


**Test program:**
	import tensorflow as tf
	import numpy as np

	batchSize=1
	numClasses=6
	numTimesteps=5

	def createGraph():
		tinputs=tf.placeholder(tf.float32, [numTimesteps, batchSize, numClasses])
		tlabels=tf.SparseTensor(tf.placeholder(tf.int64, shape=[None,2]) , tf.placeholder(tf.int32,[None]), tf.placeholder(tf.int64,[2])) # labels
		tseqLen=tf.placeholder(tf.int32, [None]) # list of sequence length in batch
		tloss=tf.reduce_mean(tf.nn.ctc_loss(labels=tlabels, inputs=tinputs, sequence_length=tseqLen, ctc_merge_repeated=True)) # ctc loss
		return (tinputs, tlabels, tseqLen, tloss)

	def getNextBatch(nc): # next batch with given number of chars in label
		indices=[[0,i] for i in range(nc)]
		values=[i%3 for i in range(nc)]
		values[0]=0
		values[1]=0 # TODO: (un)comment this to trigger warning
		shape=[1, nc]
		labels=tf.SparseTensorValue(indices, values, shape)
		seqLen=[nc]
		inputs=np.random.rand(numTimesteps, batchSize, numClasses)
		return (labels, inputs, seqLen) 


	(tinputs, tlabels, tseqLen, tloss)=createGraph()

	sess=tf.Session()
	sess.run(tf.global_variables_initializer())

	nc=3 # number of chars in label
	print('next batch with 1 element has label len='+str(nc))
	(labels, inputs, seqLen)=getNextBatch(nc)
	resLoss=sess.run([tloss], { tlabels: labels, tinputs:inputs, tseqLen:seqLen } )
	print('input: ', inputs)
	print('label: ', labels)
	print('loss: ', resLoss)

**Comment by githubharald** [comment]
ok, got it, that's not a bug, that's just how CTC works:
let's take an example for which the warning occurs: length of input sequence is 2, labelling is "aa", also with length 2.
Now the shortest path which yields "aa" is a->blank->a (length 3). But for a labelling "ab", the shortest path is a->b (length 2).
That shows why for repeated labels like in "aa" the input sequence must be longer. Its simply the way that repeated labels get encoded in the CTC (by inserting blanks).

**Comment by argman** [comment]
@vcarbune 
Note that ctc_merge_repeated is different from collapsing repeated labels, which is done through preprocess_collapse_repeated. The former controls how the CTC loss penalizes internally repeated labels ('a' 'a' 'b' becomes 'blank' 'a' 'blank' 'a' 'blank' 'b' 'blank' for the forward/backward computation), while the latter actually modifies the ground truth. ('a' 'a' 'b' first becomes 'a' 'b' and then 'blank' 'a' 'blank' 'b' 'blank' for the forward/backward computation).

I think ctc insert blank between labels as default behaviour, what does ctc_merge_repeated do?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 180 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
